File: pycti/api/opencti_api_work.py
import time
from typing import Dict, List


class OpenCTIApiWork:
    """OpenCTIApiJob"""

    def __init__(self, api):
        self.api = api

    def to_received(self, work_id: str, message: str):
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info(
                "Reporting work update_received", {"work_id": work_id}
            )
            query = """
                mutation workToReceived($id: ID!, $message: String) {
                    workEdit(id: $id) {
                        toReceived (message: $message)
                    }
                }
               """
            self.api.query(query, {"id": work_id, "message": message})

    def to_processed(self, work_id: str, message: str, in_error: bool = False):
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info(
                "Reporting work update_processed", {"work_id": work_id}
            )
            query = """
                mutation workToProcessed($id: ID!, $message: String, $inError: Boolean) {
                    workEdit(id: $id) {
                        toProcessed (message: $message, inError: $inError)
                    }
                }
               """
            self.api.query(
                query, {"id": work_id, "message": message, "inError": in_error}
            )

    def ping(self, work_id: str):
        self.api.app_logger.info("Ping work", {"work_id": work_id})
        query = """
            mutation pingWork($id: ID!) {
                workEdit(id: $id) {
                    ping
                }
            }
           """
        self.api.query(query, {"id": work_id})

    def report_expectation(self, work_id: str, error):
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info("Report expectation", {"work_id": work_id})
            query = """
                mutation reportExpectation($id: ID!, $error: WorkErrorInput) {
                    workEdit(id: $id) {
                        reportExpectation(error: $error)
                    }
                }
               """
            try:
                self.api.query(query, {"id": work_id, "error": error})
            except:
                self.api.app_logger.error("Cannot report expectation")

    def add_expectations(self, work_id: str, expectations: int):
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info(
                "Update action expectations",
                {"work_id": work_id, "expectations": expectations},
            )
            query = """
                mutation addExpectations($id: ID!, $expectations: Int) {
                    workEdit(id: $id) {
                        addExpectations(expectations: $expectations)
                    }
                }
               """
            try:
                self.api.query(query, {"id": work_id, "expectations": expectations})
            except:
                self.api.app_logger.error("Cannot report expectation")

    def add_draft_context(self, work_id: str, draft_context: str):
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info(
                "Update draft context",
                {"work_id": work_id, "draft_context": draft_context},
            )
            query = """
                mutation addDraftContext($id: ID!, $draftContext: String) {
                    workEdit(id: $id) {
                        addDraftContext(draftContext: $draftContext)
                    }
                }
               """
            try:
                self.api.query(query, {"id": work_id, "draftContext": draft_context})
            except:
                self.api.app_logger.error("Cannot report draft context")

    def initiate_work(self, connector_id: str, friendly_name: str) -> str:
        if self.api.bundle_send_to_queue:
            self.api.app_logger.info("Initiate work", {"connector_id": connector_id})
            query = """
                mutation workAdd($connectorId: String!, $friendlyName: String) {
                    workAdd(connectorId: $connectorId, friendlyName: $friendlyName) {
                      id
                    }
                }
               """
            work = self.api.query(
                query, {"connectorId": connector_id, "friendlyName": friendly_name}
            )
            return work["data"]["workAdd"]["id"]

    def delete_work(self, work_id: str):
        query = """
        mutation ConnectorWorksMutation($workId: ID!) {
            workEdit(id: $workId) {
                delete
            }
        }"""
        work = self.api.query(
            query,
            {"workId": work_id},
        )
        return work["data"]

    def wait_for_work_to_finish(self, work_id: str):
        status = ""
        cnt = 0
        while status != "complete":
            state = self.get_work(work_id=work_id)
            if len(state) > 0:
                status = state["status"]

                if state["errors"]:
                    self.api.app_logger.error(
                        "Unexpected connector error", {"state_errors": state["errors"]}
                    )
                    return ""

            time.sleep(1)
            cnt += 1

    def get_work(self, work_id: str) -> Dict:
        query = """
        query WorkQuery($id: ID!) {
            work(id: $id) {
                id
                name
                user {
                    name
                }
                timestamp
                status
                event_source_id
                received_time
                processed_time
                completed_time
                tracking {
                    import_expected_number
                    import_processed_number
                }
                messages {
                    timestamp
                    message
                    sequence
                    source
                }
                errors {
                    timestamp
                    message
                    sequence
                    source
                }
            }
        }
        """
        result = self.api.query(
            query,
            {"id": work_id},
        )
        return result["data"]["work"]

    def get_connector_works(self, connector_id: str) -> List[Dict]:
        query = """
        query ConnectorWorksQuery(
            $count: Int
            $orderBy: WorksOrdering
            $orderMode: OrderingMode
            $filters: FilterGroup
        ) {
            works(
                first: $count
                orderBy: $orderBy
                orderMode: $orderMode
                filters: $filters
            ) {
                edges {
                    node {
                        id
                        name
                        user {
                            name
                        }
                        timestamp
                        status
                        event_source_id
                        received_time
                        processed_time
                        completed_time
                        tracking {
                            import_expected_number
                            import_processed_number
                        }
                        messages {
                            timestamp
                            message
                            sequence
                            source
                        }
                        errors {
                            timestamp
                            message
                            sequence
                            source
                        }
                    }
                }
            }
        }
        """
        result = self.api.query(
            query,
            {
                "count": 50,
                "filters": {
                    "mode": "and",
                    "filters": [{"key": "connector_id", "values": [connector_id]}],
                    "filterGroups": [],
                },
            },
        )
        result = result["data"]["works"]["edges"]
        return_value = []
        for node in result:
            node = node["node"]
            return_value.append(node)

        return sorted(return_value, key=lambda i: i["timestamp"])

File: client-java/src/main/java/io/filigran/opencti/api/OpenCTIApiWork.java
package io.filigran.opencti.api;

import io.filigran.opencti.OpenCTIApiClient;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.List;
import java.util.Map;

/**
 * API module for managing OpenCTI work operations.
 *
 * @author Filigran Team
 * @since 6.9.6
 */
@Slf4j
@RequiredArgsConstructor
public class OpenCTIApiWork {

    private final OpenCTIApiClient client;

    /**
     * Initiate a new work for a connector.
     *
     * @param connectorId the connector ID
     * @param friendlyName a friendly name for the work
     * @return the work ID
     */
    @SuppressWarnings("unchecked")
    public String initiateWork(String connectorId, String friendlyName) {
        log.info("Initiating work for connector: {}", connectorId);
        
        String mutation = """
            mutation WorkAdd($connectorId: String!, $friendlyName: String) {
                workAdd(connectorId: $connectorId, friendlyName: $friendlyName) {
                    id
                }
            }
            """;
        
        Map<String, Object> result = client.query(mutation, Map.of(
            "connectorId", connectorId,
            "friendlyName", friendlyName
        ));
        
        Map<String, Object> data = (Map<String, Object>) result.get("data");
        Map<String, Object> workAdd = (Map<String, Object>) data.get("workAdd");
        return (String) workAdd.get("id");
    }

    /**
     * Update the processing count of a work.
     *
     * @param workId the work ID
     * @param processingCount the new processing count
     */
    public void updateProcessingCount(String workId, int processingCount) {
        log.info("Updating processing count for work: {} to {}", workId, processingCount);
        
        String mutation = """
            mutation WorkEdit($id: ID!, $input: [EditInput!]!) {
                workEdit(id: $id) {
                    fieldPatch(input: $input) {
                        id
                    }
                }
            }
            """;
        
        client.query(mutation, Map.of(
            "id", workId,
            "input", List.of(Map.of(
                "key", "import_expected_number",
                "value", String.valueOf(processingCount)
            ))
        ));
    }

    /**
     * Report an error for a work.
     *
     * @param workId the work ID
     * @param errorMessage the error message
     * @param sourceId optional source ID
     */
    public void reportError(String workId, String errorMessage, String sourceId) {
        log.info("Reporting error for work: {}", workId);
        
        String mutation = """
            mutation WorkAddExpectation($id: ID!, $input: WorkToProcessInput!) {
                workEdit(id: $id) {
                    addExpectation(input: $input) {
                        id
                    }
                }
            }
            """;
        
        Map<String, Object> input = new java.util.HashMap<>();
        input.put("type", "Error");
        input.put("message", errorMessage);
        if (sourceId != null) {
            input.put("source_id", sourceId);
        }
        
        client.query(mutation, Map.of("id", workId, "input", input));
    }

    /**
     * Complete a work.
     *
     * @param workId the work ID
     * @param message completion message
     * @param inError whether the work completed in error
     */
    public void complete(String workId, String message, boolean inError) {
        log.info("Completing work: {} (inError: {})", workId, inError);
        
        String mutation = """
            mutation WorkEdit($id: ID!, $message: String, $inError: Boolean) {
                workEdit(id: $id) {
                    toProcessed(message: $message, inError: $inError)
                }
            }
            """;
        
        client.query(mutation, Map.of(
            "id", workId,
            "message", message,
            "inError", inError
        ));
    }

    /**
     * Delete a work.
     *
     * @param workId the work ID
     */
    public void delete(String workId) {
        log.info("Deleting work: {}", workId);
        
        String mutation = """
            mutation WorkDelete($id: ID!) {
                workDelete(id: $id)
            }
            """;
        
        client.query(mutation, Map.of("id", workId));
    }
}

